libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.h

```diff
@@ -0,0 +1,198 @@
+#include "AuxiliaryBus.h"
+
+class AP_Invensensev2_AuxiliaryBus;
+class AP_Invensensev2_AuxiliaryBusSlave;
+
+class AP_InertialSensor_Invensensev2 : public AP_InertialSensor_Backend
+{
+    friend AP_Invensensev2_AuxiliaryBus;
+    friend AP_Invensensev2_AuxiliaryBusSlave;
+
+public:
+    virtual ~AP_InertialSensor_Invensensev2();
+
+    static AP_InertialSensor_Invensensev2 &from(AP_InertialSensor_Backend &backend) {
+        return static_cast<AP_InertialSensor_Invensensev2&>(backend);
+    }
+
+    static AP_InertialSensor_Backend *probe(AP_InertialSensor &imu,
+                                            AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev,
+                                            enum Rotation rotation = ROTATION_NONE);
+
+    static AP_InertialSensor_Backend *probe(AP_InertialSensor &imu,
+                                            AP_HAL::OwnPtr<AP_HAL::SPIDevice> dev,
+                                            enum Rotation rotation = ROTATION_NONE);
+
+    /* update accel and gyro state */
+    bool update() override;
+    void accumulate() override;
+
+    /*
+     * Return an AuxiliaryBus if the bus driver allows it
+     */
+    AuxiliaryBus *get_auxiliary_bus() override;
+
+    void start() override;
+
+    enum Invensensev2_Type {
+        Invensensev2_ICM20948 = 0,
+        Invensensev2_ICM20648,
+        Invensensev2_ICM20649
+    };
+
+    // acclerometers on Invensense sensors will return values up to
+    // 24G, but they are not guaranteed to be remotely linear past
+    // 16G
+    const uint16_t multiplier_accel = INT16_MAX/(26*GRAVITY_MSS);
+
+private:
+    AP_InertialSensor_Invensensev2(AP_InertialSensor &imu,
+                              AP_HAL::OwnPtr<AP_HAL::Device> dev,
+                              enum Rotation rotation);
+
+    /* Initialize sensor*/
+    bool _init();
+    bool _hardware_init();
+    bool _check_whoami();
+
+    void _set_filter_and_scaling(void);
+    void _fifo_reset();
+    bool _has_auxiliary_bus();
+
+    /* Read samples from FIFO (FIFO enabled) */
+    void _read_fifo();
+
+    /* Check if there's data available by either reading DRDY pin or register */
+    bool _data_ready();
+
+    /* Poll for new data (non-blocking) */
+    void _poll_data();
+
+    /* Read and write functions taking the differences between buses into
+     * account */
+    bool _block_read(uint16_t reg, uint8_t *buf, uint32_t size);
+    uint8_t _register_read(uint16_t reg);
+    void _register_write(uint16_t reg, uint8_t val, bool checked=false);
+    void _select_bank(uint8_t bank);
+
+    bool _accumulate(uint8_t *samples, uint8_t n_samples);
+    bool _accumulate_sensor_rate_sampling(uint8_t *samples, uint8_t n_samples);
+
+    bool _check_raw_temp(int16_t t2);
+
+    int16_t _raw_temp;
+    
+    // instance numbers of accel and gyro data
+    uint8_t _gyro_instance;
+    uint8_t _accel_instance;
+
+    float temp_sensitivity = 1.0/340; // degC/LSB
+    float temp_zero = 36.53; // degC
```

Comment:
did you check that temp_zero and sensitivity is the same for all these sensors? (each sensor seems to have different values for the old ones)

Reply:
Yes, they are same for all, the actual value is being set here:
https://github.com/ArduPilot/ardupilot/blob/9f50feadfed3e0066e5c5fcc0db147fcb6abcac4/libraries/AP_InertialSensor/AP_InertialSensor_Invensensev2.cpp#L206
I will update the code to be done for default values instead.